File: packages/indexer/__tests__/worker-protocol.test.ts
import { describe, expect, it } from 'vitest'
import type { ResolvedProjectModel } from '@use-crux/core/project-index'
import type { IndexPatch } from '../src/indexer/patches'
import {
  indexPatchFromWorkerEvents,
  projectIndexArtifactToWorkerEvents,
  indexPatchToWorkerEvents,
  projectIndexArtifactToWorkerEvent,
} from '../src/contracts/worker-events'
import {
  workerEventFixtureOptions,
  workerEventFixturePatch,
} from '../src/contracts/worker-events/fixtures'
import { readStaticIndexRuntimeSharedFixture } from '../src/contracts/fixtures'

describe('project index worker protocol', () => {
  it('streams contract fixture facts in ordered batches and reconstructs the same patch', () => {
    const events = indexPatchToWorkerEvents(workerEventFixturePatch, workerEventFixtureOptions)

    expect(events.map((event) => event.type)).toEqual([
      'phase:start',
      'fact:batch',
      'fact:batch',
      'sourceProfile:batch',
      'phase:done',
    ])
    expect(events[1]).toMatchObject({ type: 'fact:batch', sequence: 0 })
    expect(events[2]).toMatchObject({ type: 'fact:batch', sequence: 1 })

    expect(indexPatchFromWorkerEvents(events)).toEqual(workerEventFixturePatch)
  })

  it('reconstructs the shared worker event fixture file', () => {
    const fixture = readStaticIndexRuntimeSharedFixture('worker-events')

    expect(fixture.events.map((event) => event.type)).toEqual([
      'phase:start',
      'fact:batch',
      'sourceProfile:batch',
      'phase:done',
    ])
    expect(indexPatchFromWorkerEvents(fixture.events)).toMatchObject({
      phase: 'ast',
      project: { root: '/repo', name: 'contract-spine' },
      facts: {
        definitions: [expect.objectContaining({ id: 'prompt:contract-spine' })],
        diagnostics: [expect.objectContaining({ id: 'diagnostic:contract-spine' })],
      },
    })
  })

  it('loads shared worker event edge-case fixtures', () => {
    const fixture = readStaticIndexRuntimeSharedFixture('worker-event-cases')

    expect(fixture.artifactDone).toMatchObject({
      protocolVersion: 2,
      type: 'artifact:done',
      artifact: 'projectStaticSyntaxPlan',
      root: '/repo',
    })
    expect(fixture.artifactError).toMatchObject({
      protocolVersion: 2,
      type: 'artifact:error',
      error: { message: 'static syntax plan failed' },
    })
    expect(fixture.phaseError).toMatchObject({
      protocolVersion: 2,
      type: 'phase:error',
      phase: 'ast',
      error: { code: 'E_STATIC_INDEX' },
    })
    expect(fixture.outOfOrderEvents.map((event) => event.type)).toEqual(['phase:start', 'fact:batch'])
    expect(fixture.outOfOrderEvents[1]).toMatchObject({ type: 'fact:batch', sequence: 1 })
  })

  it('streams source profile rows and carries terminal profile metadata on phase:done', () => {
    const patch: IndexPatch = {
      schemaVersion: 1,
      phase: 'ast',
      project: { root: '/repo', name: 'fixture' },
      startedAt: '2026-06-18T10:00:00.000Z',
      finishedAt: '2026-06-18T10:00:00.001Z',
      status: 'ok',
      facts: {},
      semanticSourceProfile: {
        files: [
          {
            file: '/repo/src/writer.ts',
            sourceHash: 'hash',
            sourceBytes: 10,
            hints: { nativeDirectCruxCandidate: true, cruxCallNames: ['prompt'] },
          },
        ],
        dependencyClosure: ['/repo/src/writer.ts'],
        sourceBytes: 10,
        complete: true,
      },
    }

    const events = indexPatchToWorkerEvents(patch, {
      transactionId: 'tx-ast',
      producer: { name: '@use-crux/indexer', version: 'test' },
      maxFactsPerBatch: 2,
    })

    expect(events.map((event) => event.type)).toEqual(['phase:start', 'sourceProfile:batch', 'phase:done'])
    const semanticSourceProfile = patch.semanticSourceProfile
    if (!semanticSourceProfile) throw new Error('Expected semantic source profile fixture')
    const sourceProfileBatch = events.find((event) => event.type === 'sourceProfile:batch')
    expect(sourceProfileBatch).toMatchObject({ type: 'sourceProfile:batch', files: semanticSourceProfile.files })
    expect(sourceProfileBatch).not.toHaveProperty('complete')
    expect(events.find((event) => event.type === 'phase:done')).toMatchObject({
      patch: { semanticSourceProfile },
    })
    expect(indexPatchFromWorkerEvents(events)).toEqual(patch)
  })

  it('splits fact batches by serialized event bytes', () => {
    const patch: IndexPatch = {
      schemaVersion: 1,
      phase: 'ast',
      project: { root: '/repo', name: 'fixture' },
      startedAt: '2026-07-06T10:00:00.000Z',
      finishedAt: '2026-07-06T10:00:00.001Z',
      status: 'ok',
      facts: {
        diagnostics: [
          diagnosticFixture('diagnostic:large-a'),
          diagnosticFixture('diagnostic:large-b'),
        ],
      },
    }

    const events = indexPatchToWorkerEvents(patch, {
      transactionId: 'tx-large-facts',
      producer: { name: '@use-crux/indexer', version: 'test' },
      maxFactsPerBatch: 10,
      maxEventBytes: 1_500,
    })
    const factBatches = events.filter((event) => event.type === 'fact:batch')

    expect(factBatches).toHaveLength(2)
    expect(factBatches.map((event) => event.facts)).toEqual([[expect.any(Object)], [expect.any(Object)]])
    expect(indexPatchFromWorkerEvents(events)).toEqual(patch)
  })

  it('streams JSON artifacts through typed V2 artifact events', () => {
    const projectModel = {
      root: {
        value: '/repo',
        provenance: { kind: 'filesystem', path: '/repo', convention: 'resolved project root' },
      },
      resolutionMode: {
        value: 'config-policy',
        provenance: { kind: 'runtime', attribute: 'project-model.resolutionMode' },
      },
      configFiles: [],
      sourceRoots: [],
      ignoredPaths: [],
      definitions: [],
      relations: [],
      quality: {
        persistenceRoot: {
          value: '/repo/.crux/quality',
          provenance: {
            kind: 'filesystem',
            path: '/repo/.crux/quality',
            convention: 'default quality persistence root',
          },
        },
        includeGlobs: [],
        excludeGlobs: [],
        evaluationFiles: [],
      },
      diagnostics: [],
    } satisfies ResolvedProjectModel

    const event = projectIndexArtifactToWorkerEvent('projectModel', projectModel, {
      root: '/repo',
      transactionId: 'artifact-project-model',
    })

    expect(event).toMatchObject({
      protocolVersion: 2,
      type: 'artifact:done',
      artifact: 'projectModel',
      root: '/repo',
      payload: projectModel,
    })
  })

  it('chunks JSON artifacts by serialized event bytes', () => {
    const payload = {
      root: '/repo',
      files: [
        { file: '/repo/src/a.ts', source: 'x'.repeat(800) },
        { file: '/repo/src/b.ts', source: 'y'.repeat(800) },
      ],
    } as unknown as ResolvedProjectModel

    const events = projectIndexArtifactToWorkerEvents('projectModel', payload, {
      root: '/repo',
      transactionId: 'artifact-large-project-model',
      maxEventBytes: 700,
    })
    const chunks = events.filter((event) => event.type === 'artifact:chunk')
    const done = events.at(-1)
    const decoded = Buffer.concat(chunks.map((event) => Buffer.from(event.payloadChunk, 'base64'))).toString('utf8')

    expect(chunks.length).toBeGreaterThan(1)
    expect(done).toMatchObject({
      type: 'artifact:done',
      artifact: 'projectModel',
      root: '/repo',
    })
    expect(done).not.toHaveProperty('payload')
    expect(JSON.parse(decoded)).toEqual(payload)
  })

  it('marks runtime patch facts as runtime-observed evidence', () => {
    const patch: IndexPatch = {
      schemaVersion: 1,
      phase: 'runtime',
      project: { root: '/repo', name: 'fixture' },
      startedAt: '2026-06-20T10:00:00.000Z',
      finishedAt: '2026-06-20T10:00:00.001Z',
      status: 'ok',
      facts: {
        definitions: [
          {
            id: 'prompt:runtime',
            kind: 'prompt',
            name: 'runtime',
            fidelity: 'resolved',
            status: 'active',
          },
        ],
      },
    }

    const events = indexPatchToWorkerEvents(patch, {
      transactionId: 'tx-runtime',
      producer: { name: '@use-crux/indexer/project-runtime-indexer', version: 'test' },
    })
    const batch = events.find((event) => event.type === 'fact:batch')

    expect(batch).toMatchObject({
      type: 'fact:batch',
      facts: [
        expect.objectContaining({
          fidelity: 'runtime-observed',
          provenance: { kind: 'runtime', attribute: 'project-index.runtime' },
        }),
      ],
    })
  })
})

function diagnosticFixture(id: string): NonNullable<IndexPatch['facts']['diagnostics']>[number] {
  return {
    id,
    severity: 'info',
    code: 'index.large_fact',
    message: 'x'.repeat(800),
  }
}

File: packages/local/internal/projectindex/wire/worker_protocol_test.go
package wire

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestProjectIndexPatchStreamCollectorBuildsPatchFromOrderedBatches(t *testing.T) {
	collector := NewProjectIndexPatchStreamCollector(ProjectIndexPatchStreamOptions{
		Root:   "/repo",
		Budget: IndexPatchBudget{MaxDefinitions: 10, MaxDiagnostics: 10, MaxSources: 10, MaxBytes: 1024 * 1024},
	})

	events := []map[string]any{
		{
			"protocolVersion": 2,
			"type":            "phase:start",
			"transactionId":   "tx-ast",
			"phase":           "ast",
			"root":            "/repo",
			"startedAt":       "2026-06-18T10:00:00.000Z",
		},
		{
			"protocolVersion": 2,
			"type":            "fact:batch",
			"transactionId":   "tx-ast",
			"sequence":        0,
			"facts": []map[string]any{
				{
					"schemaVersion": 1,
					"factId":        "definitions:prompt:writer",
					"kind":          "definitions",
					"phase":         "ast",
					"projectRoot":   "/repo",
					"producer":      map[string]any{"name": "@use-crux/indexer", "version": "test"},
					"fidelity":      "inferred",
					"provenance":    map[string]any{"kind": "runtime", "attribute": "project-index.ast"},
					"fact": map[string]any{
						"id":       "prompt:writer",
						"kind":     "prompt",
						"name":     "writer",
						"fidelity": "partial",
						"status":   "active",
					},
				},
				{
					"schemaVersion": 1,
					"factId":        "diagnostics:diagnostic:writer",
					"kind":          "diagnostics",
					"phase":         "ast",
					"projectRoot":   "/repo",
					"producer":      map[string]any{"name": "@use-crux/indexer", "version": "test"},
					"fidelity":      "inferred",
					"provenance":    map[string]any{"kind": "runtime", "attribute": "project-index.ast"},
					"fact": map[string]any{
						"id":       "diagnostic:writer",
						"severity": "info",
						"code":     "index.writer",
						"message":  "writer indexed",
					},
				},
			},
		},
		{
			"protocolVersion": 2,
			"type":            "fact:batch",
			"transactionId":   "tx-ast",
			"sequence":        1,
			"facts": []map[string]any{
				{
					"schemaVersion": 1,
					"factId":        "sources:/repo/src/writer.ts",
					"kind":          "sources",
					"phase":         "ast",
					"projectRoot":   "/repo",
					"producer":      map[string]any{"name": "@use-crux/indexer", "version": "test"},
					"fidelity":      "inferred",
					"provenance":    map[string]any{"kind": "runtime", "attribute": "project-index.ast"},
					"fact": map[string]any{
						"file":          "/repo/src/writer.ts",
						"status":        "active",
						"shardId":       ".",
						"definitionIds": []string{"prompt:writer"},
						"diagnostics":   []string{"diagnostic:writer"},
					},
				},
				{
					"schemaVersion": 1,
					"factId":        "sourceGraph:0",
					"kind":          "sourceGraph",
					"phase":         "ast",
					"projectRoot":   "/repo",
					"producer":      map[string]any{"name": "@use-crux/indexer", "version": "test"},
					"fidelity":      "inferred",
					"provenance":    map[string]any{"kind": "runtime", "attribute": "project-index.ast"},
					"fact": map[string]any{
						"schemaVersion": 1,
						"producedBy":    "@use-crux/indexer",
						"capabilities":  []string{"project-shards"},
						"shards": []map[string]any{
							{"id": ".", "root": "/repo", "packageFile": "/repo/package.json"},
						},
					},
				},
			},
		},
		{
			"protocolVersion": 2,
			"type":            "phase:done",
			"transactionId":   "tx-ast",
			"phase":           "ast",
			"patch": map[string]any{
				"schemaVersion": 1,
				"phase":         "ast",
				"project":       map[string]any{"root": "/repo", "name": "fixture"},
				"startedAt":     "2026-06-18T10:00:00.000Z",
				"finishedAt":    "2026-06-18T10:00:00.001Z",
				"status":        "ok",
				"invalidates":   map[string]any{"all": true},
			},
			"summary": map[string]any{
				"factCount": 4,
				"timings": []map[string]any{
					{"name": "static.cache.read", "durationMs": 12.5, "count": 2},
				},
			},
		},
	}

	for _, event := range events {
		if err := collector.Handle(mustMarshalWorkerEvent(t, event)); err != nil {
			t.Fatalf("Handle(%s) error = %v", event["type"], err)
		}
	}

	patches, err := collector.Patches()
	if err != nil {
		t.Fatalf("Patches error = %v", err)
	}
	if len(patches) != 1 {
		t.Fatalf("patches len = %d, want 1", len(patches))
	}
	patch := patches[0]
	if patch.Phase != PhaseAST || patch.Project.Root != "/repo" || patch.Status != "ok" {
		t.Fatalf("patch metadata = %+v", patch)
	}
	if len(patch.Facts.Definitions) != 1 || patch.Facts.Definitions[0].ID != "prompt:writer" {
		t.Fatalf("definitions = %+v, want prompt:writer", patch.Facts.Definitions)
	}
	if len(patch.Facts.Diagnostics) != 1 || patch.Facts.Diagnostics[0].Code != "index.writer" {
		t.Fatalf("diagnostics = %+v, want index.writer", patch.Facts.Diagnostics)
	}
	if len(patch.Facts.Sources) != 1 || patch.Facts.Sources[0].File != "/repo/src/writer.ts" {
		t.Fatalf("sources = %+v, want writer source", patch.Facts.Sources)
	}
	if patch.Facts.Sources[0].ShardID != "." {
		t.Fatalf("source shardId = %q, want .", patch.Facts.Sources[0].ShardID)
	}
	if patch.Facts.SourceGraph == nil || len(patch.Facts.SourceGraph.Shards) != 1 || patch.Facts.SourceGraph.Shards[0].ID != "." {
		t.Fatalf("sourceGraph = %+v, want root shard", patch.Facts.SourceGraph)
	}
	timings := collector.Timings()
	if len(timings) != 1 || timings[0].Name != "static.cache.read" || timings[0].DurationMs != 12.5 || timings[0].Count != 2 {
		t.Fatalf("timings = %+v, want static.cache.read summary", timings)
	}
}

func TestProjectIndexPatchStreamCollectorBuildsSemanticSourceProfileFromBatches(t *testing.T) {
	collector := NewProjectIndexPatchStreamCollector(ProjectIndexPatchStreamOptions{Root: "/repo"})

	events := []map[string]any{
		{
			"protocolVersion": 2,
			"type":            "phase:start",
			"transactionId":   "tx-ast",
			"phase":           "ast",
			"root":            "/repo",
			"startedAt":       "2026-06-18T10:00:00.000Z",
		},
		{
			"protocolVersion": 2,
			"type":            "sourceProfile:batch",
			"transactionId":   "tx-ast",
			"sequence":        0,
			"files": []map[string]any{
				{
					"file":        "/repo/src/a.ts",
					"sourceHash":  "hash-a",
					"sourceBytes": 12,
					"hints": map[string]any{
						"nativeDirectCruxCandidate": true,
						"cruxCallNames":             []string{"prompt"},
					},
				},
			},
		},
		{
			"protocolVersion": 2,
			"type":            "phase:done",
			"transactionId":   "tx-ast",
			"phase":           "ast",
			"patch": map[string]any{
				"schemaVersion": 1,
				"phase":         "ast",
				"project":       map[string]any{"root": "/repo"},
				"startedAt":     "2026-06-18T10:00:00.000Z",
				"finishedAt":    "2026-06-18T10:00:00.001Z",
				"status":        "ok",
			},
			"summary": map[string]any{"factCount": 0},
		},
	}

	for _, event := range events {
		if err := collector.Handle(mustMarshalWorkerEvent(t, event)); err != nil {
			t.Fatalf("Handle(%s) error = %v", event["type"], err)
		}
	}
	patches, err := collector.Patches()
	if err != nil {
		t.Fatalf("Patches error = %v", err)
	}
	profile := patches[0].SemanticSourceProfile
	if profile == nil || len(profile.Files) != 1 {
		t.Fatalf("semantic source profile = %+v, want streamed profile file", profile)
	}
	if profile.SourceBytes != 12 || !profile.Complete {
		t.Fatalf("semantic source profile = %+v, want byte count and complete marker", profile)
	}
	if profile.Files[0].Hints == nil || !profile.Files[0].Hints.NativeDirectCruxCandidate {
		t.Fatalf("semantic source profile hints = %+v, want native candidate", profile.Files[0].Hints)
	}
}

func TestProjectIndexPatchStreamCollectorRejectsSequenceGaps(t *testing.T) {
	collector := NewProjectIndexPatchStreamCollector(ProjectIndexPatchStreamOptions{Root: "/repo"})

	if err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "phase:start",
		"transactionId":   "tx-ast",
		"phase":           "ast",
		"root":            "/repo",
		"startedAt":       "2026-06-18T10:00:00.000Z",
	})); err != nil {
		t.Fatalf("Handle phase:start error = %v", err)
	}

	err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "fact:batch",
		"transactionId":   "tx-ast",
		"sequence":        1,
		"facts":           []map[string]any{},
	}))

	if err == nil || !strings.Contains(err.Error(), "sequence") {
		t.Fatalf("Handle sequence gap error = %v, want sequence error", err)
	}
}

func TestProjectIndexPatchStreamCollectorRejectsRootMismatch(t *testing.T) {
	collector := NewProjectIndexPatchStreamCollector(ProjectIndexPatchStreamOptions{Root: "/repo"})

	err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "phase:start",
		"transactionId":   "tx-ast",
		"phase":           "ast",
		"root":            "/other",
		"startedAt":       "2026-06-18T10:00:00.000Z",
	}))

	if err == nil || !strings.Contains(err.Error(), "root") {
		t.Fatalf("Handle root mismatch error = %v, want root error", err)
	}
}

func TestProjectIndexPatchStreamCollectorRejectsPatchBudgetViolations(t *testing.T) {
	collector := NewProjectIndexPatchStreamCollector(ProjectIndexPatchStreamOptions{
		Root:   "/repo",
		Budget: IndexPatchBudget{MaxDefinitions: 1},
	})

	events := []map[string]any{
		{
			"protocolVersion": 2,
			"type":            "phase:start",
			"transactionId":   "tx-ast",
			"phase":           "ast",
			"root":            "/repo",
			"startedAt":       "2026-06-18T10:00:00.000Z",
		},
		{
			"protocolVersion": 2,
			"type":            "fact:batch",
			"transactionId":   "tx-ast",
			"sequence":        0,
			"facts": []map[string]any{
				testDefinitionFact("prompt:one"),
				testDefinitionFact("prompt:two"),
			},
		},
	}
	for _, event := range events {
		if err := collector.Handle(mustMarshalWorkerEvent(t, event)); err != nil {
			t.Fatalf("Handle(%s) error = %v", event["type"], err)
		}
	}

	err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "phase:done",
		"transactionId":   "tx-ast",
		"phase":           "ast",
		"patch": map[string]any{
			"schemaVersion": 1,
			"phase":         "ast",
			"project":       map[string]any{"root": "/repo"},
			"startedAt":     "2026-06-18T10:00:00.000Z",
			"finishedAt":    "2026-06-18T10:00:00.001Z",
			"status":        "ok",
		},
		"summary": map[string]any{"factCount": 2},
	}))

	if err == nil || !strings.Contains(err.Error(), "definitions 2/1") {
		t.Fatalf("Handle budget violation error = %v, want definitions budget error", err)
	}
}

func TestProjectIndexPatchStreamCollectorRejectsProducerMismatch(t *testing.T) {
	collector := NewProjectIndexPatchStreamCollector(ProjectIndexPatchStreamOptions{
		Root:     "/repo",
		Producer: "@use-crux/indexer/project-indexer",
	})

	if err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "phase:start",
		"transactionId":   "tx-ast",
		"phase":           "ast",
		"root":            "/repo",
		"startedAt":       "2026-06-18T10:00:00.000Z",
	})); err != nil {
		t.Fatalf("Handle phase:start error = %v", err)
	}

	err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "fact:batch",
		"transactionId":   "tx-ast",
		"sequence":        0,
		"facts":           []map[string]any{testDefinitionFact("prompt:writer")},
	}))

	if err == nil || !strings.Contains(err.Error(), "producer") {
		t.Fatalf("Handle producer mismatch error = %v, want producer error", err)
	}
}

func TestProjectIndexPatchStreamCollectorPreservesRuntimeObservedEnvelopeMetadata(t *testing.T) {
	collector := NewProjectIndexPatchStreamCollector(ProjectIndexPatchStreamOptions{Root: "/repo"})
	if err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "phase:start",
		"transactionId":   "tx-runtime",
		"phase":           "runtime",
		"root":            "/repo",
		"startedAt":       "2026-06-20T10:00:00.000Z",
	})); err != nil {
		t.Fatalf("Handle phase:start error = %v", err)
	}
	if err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "fact:batch",
		"transactionId":   "tx-runtime",
		"sequence":        0,
		"facts": []map[string]any{{
			"schemaVersion": 1,
			"factId":        "definitions:prompt:runtime",
			"kind":          "definitions",
			"phase":         "runtime",
			"projectRoot":   "/repo",
			"producer":      map[string]any{"name": "@use-crux/indexer/project-runtime-indexer", "version": "test"},
			"fidelity":      "runtime-observed",
			"provenance":    map[string]any{"kind": "runtime", "attribute": "project-index.runtime"},
			"fact": map[string]any{
				"id":       "prompt:runtime",
				"kind":     "prompt",
				"name":     "runtime",
				"fidelity": "resolved",
				"status":   "active",
			},
		}},
	})); err != nil {
		t.Fatalf("Handle fact:batch error = %v", err)
	}
	if err := collector.Handle(mustMarshalWorkerEvent(t, map[string]any{
		"protocolVersion": 2,
		"type":            "phase:done",
		"transactionId":   "tx-runtime",
		"phase":           "runtime",
		"patch": map[string]any{
			"schemaVersion": 1,
			"phase":         "runtime",
			"project":       map[string]any{"root": "/repo"},
			"startedAt":     "2026-06-20T10:00:00.000Z",
			"finishedAt":    "2026-06-20T10:00:00.001Z",
			"status":        "ok",
		},
		"summary": map[string]any{"factCount": 1},
	})); err != nil {
		t.Fatalf("Handle phase:done error = %v", err)
	}

	patches, err := collector.Patches()
	if err != nil {
		t.Fatalf("Patches error = %v", err)
	}
	envelopes := patches[0].FactEnvelopes
	if len(envelopes) != 1 {
		t.Fatalf("fact envelopes = %+v, want one runtime envelope", envelopes)
	}
	if envelopes[0].Fidelity != "runtime-observed" || envelopes[0].Provenance.Attribute != "project-index.runtime" {
		t.Fatalf("runtime envelope metadata = %+v, want runtime observed provenance", envelopes[0])
	}
}

func testDefinitionFact(id string) map[string]any {
	return map[string]any{
		"schemaVersion": 1,
		"factId":        "definitions:" + id,
		"kind":          "definitions",
		"phase":         "ast",
		"projectRoot":   "/repo",
		"producer":      map[string]any{"name": "@use-crux/indexer", "version": "test"},
		"fidelity":      "inferred",
		"provenance":    map[string]any{"kind": "runtime", "attribute": "project-index.ast"},
		"fact": map[string]any{
			"id":       id,
			"kind":     "prompt",
			"name":     id,
			"fidelity": "partial",
			"status":   "active",
		},
	}
}

func mustMarshalWorkerEvent(t *testing.T, value any) json.RawMessage {
	t.Helper()
	data, err := json.Marshal(value)
	if err != nil {
		t.Fatalf("marshal event: %v", err)
	}
	return data
}
